### ISSUE TITLE
Help. How to bring in libraries with transitive dependecies?

### ISSUE THREAD

**Issue by RafaGago** [tweag/rules_nixpkgs#141]
Hello,

I'm kind of new on both Bazel and Nix, so I only know the basics.

I was toying with doing a simple GTK hello world using this project to see how viable is to use Nix with this integration at work.

I did chose to do a GTK helloworld because GTK is a library with lots of dependencies. The existing examples with zlib are just the happy case, where the package has no dependencies on itself.

What I did so far is on [this repository](https://github.com/RafaGago/nix-bazel-gtk-helloworld). I basically imported gtk and all the required dependencies appearing on the headers.

As NIX is a package manager, I see that all the dependencies of GTK are correctly brought to the store, but it seems that on the Bazel side some of the dependencies need to be replicated/Bazelified. 

So e.g. "gtk-3.dev" doesn't bring in all the ".dev" packages of its dependencies containing extra headers, I had to manually import every dependency appearing on the headers and to fix the include paths, including version numbers (this might be solved by pkgconfig, but I did it manually for testing purposes).

I was expecting that after doing so I would be able to link. This is the content of "k8-fastbuild/bin/src/gtkhello-2.params" showing that the depencies are passed to the linker.

```
-o
bazel-out/k8-fastbuild/bin/src/gtkhello
-Wl,-rpath,$ORIGIN/../_solib_k8/_U@gtk3-x11_S_S_Cgtk3-so___Uexternal_Sgtk3-x11_Slib_Sgtk-3.0_S3.0.0_Simmodules
-Wl,-rpath,$ORIGIN/../_solib_k8/_U@gtk3-x11_S_S_Cgtk3-so___Uexternal_Sgtk3-x11_Slib_Sgtk-3.0_S3.0.0_Sprintbackends
-Wl,-rpath,$ORIGIN/../_solib_k8/_U@gtk3-x11_S_S_Cgtk3-so___Uexternal_Sgtk3-x11_Slib
-Wl,-rpath,$ORIGIN/../_solib_k8/_U@glib.out_S_S_Cglib-so___Uexternal_Sglib.out_Slib
-Lbazel-out/k8-fastbuild/bin/_solib_k8/_U@gtk3-x11_S_S_Cgtk3-so___Uexternal_Sgtk3-x11_Slib_Sgtk-3.0_S3.0.0_Simmodules
-Lbazel-out/k8-fastbuild/bin/_solib_k8/_U@gtk3-x11_S_S_Cgtk3-so___Uexternal_Sgtk3-x11_Slib_Sgtk-3.0_S3.0.0_Sprintbackends
-Lbazel-out/k8-fastbuild/bin/_solib_k8/_U@gtk3-x11_S_S_Cgtk3-so___Uexternal_Sgtk3-x11_Slib
-Lbazel-out/k8-fastbuild/bin/_solib_k8/_U@glib.out_S_S_Cglib-so___Uexternal_Sglib.out_Slib
-Wl,-S
-fuse-ld=gold
-Wl,-no-as-needed
-Wl,-z,relro,-z,now
-B/usr/bin
-pass-exit-codes
-lstdc++
-lm
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o
-l:im-am-et.so
-l:im-cedilla.so
-l:im-cyrillic-translit.so
-l:im-inuktitut.so
-l:im-ipa.so
-l:im-multipress.so
-l:im-thai.so
-l:im-ti-er.so
-l:im-ti-et.so
-l:im-viqr.so
-l:im-wayland.so
-l:im-waylandgtk.so
-l:im-xim.so
-lprintbackend-cups
-lprintbackend-file
-lprintbackend-lpr
-lgailutil-3
-lgdk-3
-lgtk-3
-lglib-2.0
```
But this isn't working. I get missing symbols on glib:

```
Linking of rule '//src:gtkhello' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-fastbuild/bin/src/gtkhello-2.params

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o:gtkmain.c:function activate: error: undefined reference to 'g_type_check_instance_cast'
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o:gtkmain.c:function main: error: undefined reference to 'g_signal_connect_data'
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o:gtkmain.c:function main: error: undefined reference to 'g_application_get_type'
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o:gtkmain.c:function main: error: undefined reference to 'g_type_check_instance_cast'
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o:gtkmain.c:function main: error: undefined reference to 'g_application_run'
bazel-out/k8-fastbuild/bin/src/_objs/gtkhello/gtkmain.pic.o:gtkmain.c:function main: error: undefined reference to 'g_object_unref'
collect2: error: ld returned 1 exit status
Target //src:gtkhello failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.173s, Critical Path: 0.07s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

Am I missing something? is this the way to go?

**Comment by aherrmann** [comment]
Hi @RafaGago,

Thanks for raising this issue. You're right that this use-case is not documented so far and that `rules_nixpkgs` does not automatically capture transitive dependencies in the way that's required here.

I looked at your repository and I think you are on the right track.

I was able to reproduce your issue with the following command (pinning nixpkgs to the latest nixpkgs-unstable to use Bazel version 3.3.1):
```
NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/1d8018068278a717771e9ec4054dff1ebd3252b0.tar.gz nix-shell -p bazel --run 'bazel build //src:gtkhello -s'
```

The linker error points at missing symbols from the libraries `libgio-2.0.so` and `libgobject-2.0.so`. Your example is referencing these functions directly, so it makes sense to add them as direct dependencies to the corresponding Bazel target. With the following minimal patch I was able to get your example to compile:
``` diff
diff --git a/WORKSPACE b/WORKSPACE
index d995d59..522d000 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -65,11 +65,11 @@ cc_library(
 )
 
 # I don't use the libraries directly, letting NIX solve this.
-#cc_library(
-#    name = "gio-so",
-#    srcs = glob(["lib/libgio*.so"], allow_empty = False),
-#)
-#
+cc_library(
+    name = "gio-so",
+    srcs = glob(["lib/libgio*.so"], allow_empty = False),
+)
+
 #cc_library(
 #    name = "glib-so",
 #    srcs = glob(["lib/libglib*.so"], allow_empty = False),
@@ -80,12 +80,12 @@ cc_library(
 #    name = "gmodule-so",
 #    srcs = glob(["lib/libgmodule*.so"], allow_empty = False),
 #)
-#
-#cc_library(
-#    name = "gobject-so",
-#    srcs = glob(["lib/libgobject*.so"], allow_empty = False),
-#)
-#
+
+cc_library(
+    name = "gobject-so",
+    srcs = glob(["lib/libgobject*.so"], allow_empty = False),
+)
+
 #cc_library(
 #    name = "gthread-so",
 #    srcs = glob(["lib/libgthread*.so"], allow_empty = False),
diff --git a/src/BUILD b/src/BUILD
index 57b7fca..0b4dce1 100644
--- a/src/BUILD
+++ b/src/BUILD
@@ -3,5 +3,9 @@ load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_import")
 cc_binary(
     name = "gtkhello",
     srcs = ["gtkmain.c"],
-    deps = ["@gtk3-x11//:gtk3-so"],
+    deps = [
+        "@glib.out//:gio-so",
+        "@glib.out//:gobject-so",
+        "@gtk3-x11//:gtk3-so",
+    ],
 )
```

I'm not terribly familiar with the GTK library, so I cannot say how to best map its Nix packages to Bazel targets. However, [the Bazel documentation](https://docs.bazel.build/versions/master/build-ref.html#dependencies) provides some guiding principles around direct and transitive dependencies.

**Comment by RafaGago** [comment]
Thanks!

It seems that I got confused because the example only includes "gtk/gtk.h", so I was expecting gtk to link against those glib targets itself. As you say those might be Gtk intricacies. I don't know the library either, I just picked something random that I knew it would be nice to show the power of this if it worked from scratch, or to get an idea of what's required if it didn't.

It's still amazing and solves reproducibility better and with less maintenance than a local repository of binary packages (as it is source based).

I still have things to learn, but I think that most are out of the scope of this project. So I close this issue.
